containers/src/Data/Map/Internal.hs

```diff
@@ -4118,6 +4119,30 @@ deleteFindMax t = case maxViewWithKey t of
   Nothing -> (error "Map.deleteFindMax: can not return the maximal element of an empty map", Tip)
   Just res -> res
 
+{--------------------------------------------------------------------
+  Iterator
+--------------------------------------------------------------------}
+
+newtype Iterator k a = Iterator (Stack k a)
+
+iterDown :: Map k a -> Stack k a -> Stack k a
+iterDown Tip stk = stk
+iterDown (Bin _ kx x l r) stk = iterDown l (Push kx x r stk)
+
+iterator :: Map k a -> Iterator k a
+iterator m = Iterator (iterDown m Nada)
+
+iterNext :: Iterator k a -> Maybe (StrictPair (k, a) (Iterator k a))
+iterNext (Iterator stk) = case stk of
+  Nada -> Nothing
+  Push kx x r stk' -> Just $! (kx, x) :*: Iterator (iterDown r stk')
+{-# INLINE iterNext #-}
```

Comment:
Do the strict pairs go away? Do the inner pairs? Does GHC hang on to the fact that the keys are in WHNF?

Reply:
> Do the strict pairs go away? Do the inner pairs?

Yes, this is inlined so it goes away.

> Does GHC hang on to the fact that the keys are in WHNF?

This code is a little obsolete, I changed `(kx, x)` to `KeyValue kx x` so it should keep track of that. But since it's inlined I would expect it to do so anyway since the tuple will be gone and `Push` is strict in the key.

Reply:
Here's the Core for

```hs
foo :: Map Int Int -> Map Int Int -> Ordering
foo = compare
```

On GHC 9.2.8 (decided to check a slightly older GHC)

<details>

```hs
Rec {
-- RHS size: {terms: 149, types: 164, coercions: 0, joins: 0/0}
$wgo13 [InlPrag=[2], Occ=LoopBreaker]
  :: Map Int Int -> Stack Int Int -> (# Ordering, Stack Int Int #)
[GblId, Arity=2, Str=<1L><1L>, Unf=OtherCon []]
$wgo13
  = \ (w_s6Ea :: Map Int Int) (w1_s6Eb :: Stack Int Int) ->
      case w_s6Ea of {
        Bin dt_a4XK k1_a4XL v2_a4XM ds1_a4XN ds2_a4XO ->
          case k1_a4XL of { I# x#_s6Wl ->
          case dt_a4XK of {
            __DEFAULT ->
              case $wgo13 ds1_a4XN w1_s6Eb of ww_s6Ee
              { (# ww1_s6U9, ww2_s6Ua #) ->
              case ww1_s6U9 of {
                __DEFAULT -> ww_s6Ee;
                EQ ->
                  case ww2_s6Ua of {
                    Push kx1_a4WQ x1_a4WR r_a4WS stk_a4WT ->
                      case kx1_a4WQ of { I# y#_s6Wo ->
                      case iterDown @Int @Int r_a4WS stk_a4WT of dt1_a4WV { __DEFAULT ->
                      case <# x#_s6Wl y#_s6Wo of {
                        __DEFAULT ->
                          case ==# x#_s6Wl y#_s6Wo of {
                            __DEFAULT -> (# GT, dt1_a4WV #);
                            1# ->
                              case v2_a4XM of { I# x#1_X1F ->
                              case x1_a4WR of { I# y#1_X1H ->
                              case <# x#1_X1F y#1_X1H of {
                                __DEFAULT ->
                                  case ==# x#1_X1F y#1_X1H of {
                                    __DEFAULT -> (# GT, dt1_a4WV #);
                                    1# -> $wgo13 ds2_a4XO dt1_a4WV
                                  };
                                1# -> (# LT, dt1_a4WV #)
                              }
                              }
                              }
                          };
                        1# -> (# LT, dt1_a4WV #)
                      }
                      }
                      };
                    Nada -> (# GT, Nada @Int @Int #)
                  }
              }
              };
            1# ->
              case w1_s6Eb of {
                Push kx1_a4WQ x1_a4WR r_a4WS stk_a4WT ->
                  case kx1_a4WQ of { I# y#_s6Wu ->
                  case iterDown @Int @Int r_a4WS stk_a4WT of dt1_a4WV { __DEFAULT ->
                  case <# x#_s6Wl y#_s6Wu of {
                    __DEFAULT ->
                      case ==# x#_s6Wl y#_s6Wu of {
                        __DEFAULT -> (# GT, dt1_a4WV #);
                        1# ->
                          case v2_a4XM of { I# x#1_X1F ->
                          case x1_a4WR of { I# y#1_X1H ->
                          case <# x#1_X1F y#1_X1H of {
                            __DEFAULT ->
                              case ==# x#1_X1F y#1_X1H of {
                                __DEFAULT -> (# GT, dt1_a4WV #);
                                1# -> (# EQ, dt1_a4WV #)
                              };
                            1# -> (# LT, dt1_a4WV #)
                          }
                          }
                          }
                      };
                    1# -> (# LT, dt1_a4WV #)
                  }
                  }
                  };
                Nada -> (# GT, Nada @Int @Int #)
              }
          }
          };
        Tip -> case w1_s6Eb of dt_a6ue { __DEFAULT -> (# EQ, dt_a6ue #) }
      }
end Rec }

-- RHS size: {terms: 10, types: 21, coercions: 0, joins: 0/0}
foo1 [InlPrag=[2]]
  :: Map Int Int
     -> Stack Int Int -> StrictPair Ordering (Stack Int Int)
[GblId,
 Arity=2,
 Str=<1L><1L>,
 Cpr=1,
 Unf=Unf{Src=InlineStable, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=2,unsat_ok=True,boring_ok=False)
         Tmpl= \ (w_s6Ea [Occ=Once1] :: Map Int Int)
                 (w1_s6Eb [Occ=Once1] :: Stack Int Int) ->
                 case $wgo13 w_s6Ea w1_s6Eb of
                 { (# ww1_s6U9 [Occ=Once1], ww2_s6Ua [Occ=Once1] #) ->
                 :*: @Ordering @(Stack Int Int) ww1_s6U9 ww2_s6Ua
                 }}]
foo1
  = \ (w_s6Ea :: Map Int Int) (w1_s6Eb :: Stack Int Int) ->
      case $wgo13 w_s6Ea w1_s6Eb of { (# ww1_s6U9, ww2_s6Ua #) ->
      :*: @Ordering @(Stack Int Int) ww1_s6U9 ww2_s6Ua
      }

-- RHS size: {terms: 20, types: 33, coercions: 0, joins: 0/0}
foo :: Map Int Int -> Map Int Int -> Ordering
[GblId,
 Arity=2,
 Str=<1L><1L>,
 Unf=Unf{Src=InlineStable, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=2,unsat_ok=True,boring_ok=False)
         Tmpl= \ (m1_a4Ws [Occ=Once1] :: Map Int Int)
                 (m2_a4Wt [Occ=Once1] :: Map Int Int) ->
                 case foo1 m1_a4Ws (iterDown @Int @Int m2_a4Wt (Nada @Int @Int)) of
                 { :*: o_a4Ya [Occ=Once1!] it_a4Yb [Occ=Once1!] ->
                 case o_a4Ya of {
                   LT -> LT;
                   EQ ->
                     case it_a4Yb of {
                       Push _ [Occ=Dead] _ [Occ=Dead] _ [Occ=Dead] _ [Occ=Dead] -> LT;
                       Nada -> EQ
                     };
                   GT -> GT
                 }
                 }}]
foo
  = \ (m1_a4Ws :: Map Int Int) (m2_a4Wt :: Map Int Int) ->
      case $wgo13 m1_a4Ws (iterDown @Int @Int m2_a4Wt (Nada @Int @Int))
      of
      { (# ww1_s6U9, ww2_s6Ua #) ->
      case ww1_s6U9 of wild1_a4Yd {
        __DEFAULT -> wild1_a4Yd;
        EQ ->
          case ww2_s6Ua of {
            Push ds_a5TP ds1_a5TQ ds2_a5TR ds3_a5TS -> LT;
            Nada -> EQ
          }
      }
      }
```

</details>